include/lo2s/time/time.hpp

```diff
@@ -26,15 +26,99 @@
 #include <otf2xx/chrono/chrono.hpp>
 
 #include <chrono>
+#include <ctime>
 
 #include <cstdint>
 
+#include <stdexcept> // for use by ClockProvider::InvalidClock
+#include <string>    // for use by ClockProvider::get_clock_by_name
+
 // All the time stuff is based on the assumption that all times are nanoseconds.
 namespace lo2s
 {
 namespace time
 {
-using Clock = std::chrono::steady_clock;
+
+struct Clock
+{
+    // This is except from the type = CamelCase naming convention to be compatible
+    using duration = std::chrono::nanoseconds;
+    using rep = duration::rep;
+    using period = duration::period;
+    using time_point = std::chrono::time_point<Clock, duration>;
+
+    static time_point now() noexcept
+    {
+        timespec tp;
+        clock_gettime(clockid_, &tp);
+        return time_point(duration(std::chrono::seconds(tp.tv_sec)) +
+                          duration(std::chrono::nanoseconds(tp.tv_nsec)));
+    }
+
+    static void set_clock(clockid_t id)
+    {
+        clockid_ = id;
+    }
+
+private:
+    static clockid_t clockid_;
+};
+
+struct ClockDescription
+{
+    const char* name;
+    clockid_t id;
+};
+
+class ClockProvider
+{
+public:
+    struct InvalidClock : std::runtime_error
+    {
+        InvalidClock(const std::string& what) : std::runtime_error(what)
+        {
+        }
+    };
+
+    static const ClockDescription& get_clock_by_name(const std::string& name);
+
+private:
+    static constexpr ClockDescription clocks_[] = {
+        {
+            "realtime", CLOCK_REALTIME,
+        },
+        {
+            "monotonic", CLOCK_MONOTONIC,
+        },
+        {
+            "process-cputime-id", CLOCK_PROCESS_CPUTIME_ID,
+        },
+        {
+            "process-thread-id", CLOCK_THREAD_CPUTIME_ID,
+        },
+        {
+            "monotonic-raw", CLOCK_MONOTONIC_RAW,
+        },
+        {
+            "realtime-coarse", CLOCK_REALTIME_COARSE,
+        },
+        {
+            "monotonic-coarse", CLOCK_MONOTONIC_COARSE,
+        },
+        {
+            "boottime", CLOCK_BOOTTIME,
+        },
+        {
+            "realtime-alarm", CLOCK_REALTIME_ALARM,
+        },
+        {
+            "boottime-alarm", CLOCK_BOOTTIME_ALARM,
+        },
+        {
+            "tai", CLOCK_TAI,
```

Comment:
Aaaaand there's systems where this isn't defined ಠ_ಠ (looking at you Travis).

We should probably only include clocks that are defined.

Reply:
Why doesn't this help:

> The available timers are defined in linux/time.h

Reply:
I remembered something: Travis uses very old linux header files, because Travis. And because it's terrible.
Therefore it doesn't even build with USE_PERF_CLOCKID=ON

So this might be the issue there.

Reply:
So including `linux/time.h` still doesn't work (it clashes with `ctime`). But [POSIX](http://pubs.opengroup.org/onlinepubs/009695399/basedefs/time.h.html) defines some of the `CLOCK_*` macros. I would really like a cleaner solution to this, but right now it seems like we have to rely on both headers defining the same values and continue to include from `ctime`.